crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2579 @@
+    if (bufbn == NULL) {
+        return 0;
+    }
+    cres = BN_cmp(bufbn, gorder);
+    *res = cres;
+    BN_free(bufbn);
+    BN_free(gorder);
+    return 1;
+}
+
+/*
+ * @brief generate a key pair keeping private inside API
+ *
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @param mode is the mode (currently unused)
+ * @param suite is the ciphersuite
+ * @param publen is the size of the public key buffer (exact length on output)
+ * @param pub is the public value
+ * @param priv is the private key pointer
+ * @return 1 for good (OpenSSL style), not 1 for error
+ */
+static int hpke_kg_evp(OSSL_LIB_CTX *libctx, const char *propq,
+                       unsigned int mode, OSSL_HPKE_SUITE suite,
+                       size_t ikmlen, const unsigned char *ikm,
+                       size_t *publen, unsigned char *pub,
+                       EVP_PKEY **priv)
+{
+    int erv = 1; /* Our error return value - 1 is success */
+    EVP_PKEY_CTX *pctx = NULL;
+    EVP_PKEY *skR = NULL;
+    unsigned char *lpub = NULL;
+    size_t lpublen = 0;
+    uint16_t kem_ind = 0;
+    int cmp = 0;
+    unsigned char suitebuf[2];
+    EVP_KDF_CTX *kctx = NULL;
+    const char *mdname = NULL;
+
+    if (hpke_suite_check(suite) != 1)
+        return 0;
+    if (pub == NULL || publen == NULL || *publen == 0 || priv == NULL)
+        return 0;
+    if (ikmlen > 0 && ikm == NULL)
+        return 0;
+    if (ikmlen == 0 && ikm != NULL)
+        return 0;
+    kem_ind = kem_iana2index(suite.kem_id);
+    if (kem_ind == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    mdname = hpke_kem_tab[kem_ind].mdname;
+    kctx = ossl_kdf_ctx_create("HKDF", mdname, libctx, propq);
+    if (kctx == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    suitebuf[0] = suite.kem_id / 256;
+    suitebuf[1] = suite.kem_id % 256;
+    /* setup generation of key pair */
+    if (hpke_kem_id_nist_curve(suite.kem_id) == 1) {
+        /* TODO: check this use of propq!!! */
+        pctx = EVP_PKEY_CTX_new_from_name(libctx, hpke_kem_tab[kem_ind].keytype,
+                                          (propq != NULL ? propq
+                                           : hpke_kem_tab[kem_ind].groupname)
+                                          );
+        if (pctx == NULL
+            || EVP_PKEY_paramgen_init(pctx) != 1
+            || EVP_PKEY_keygen_init(pctx) <= 0) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx,
+                                                   hpke_kem_tab[kem_ind].groupid
+                                                   ) != 1) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        if (ikm != NULL) {
```

Comment:
This code is not supposed to be here I think? The keygen can do it..

Reply:
Think it's needed - the NIST curve stuff in 9180 is... baroque and different (meaning the ``ikm != NULL`` branch) if you meant something else not sure...

Reply:
Yes it is included..

````


    OSSL_PARAM params[3];

    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME,
                                                 "secp256k1", 0);
    params[1] = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_DHKEM_IKM,
                                                  (char *)t->ikm, t->ikmlen);
    params[2] = OSSL_PARAM_construct_end();

    genctx = EVP_PKEY_CTX_new_from_name(libctx, "EC", NULL);
    EVP_PKEY_keygen_init(genctx);
    EVP_PKEY_CTX_set_params(genctx, params);
    EVP_PKEY_generate(genctx, &pkey)
````

Reply:
ECX is similar - It just doesnt have a group name..

Reply:
This means you can get rid of most of  the dependancies (including the strings at the top of the file)

Reply:
ah cool, will drop that so - did you include the test vector that increments the counter in your test stuff? be a shame to lose that if not

Reply:
Cant remember if it does.. The coverage tests hit every branch I could though.. And the code does handle it.

Reply:
turns out it works fine - the ``ikmiter[]`` test vector in ``test/test-hpke.c`` exercises that code and works ok (just checked in gdb)